commit security tutorial for review (#258)
trilitech/tezos-developer-docs: docs/tutorials/security/part-3.md

---
title: 'Part 3: User trust & management'
authors: 'Benjamin Fuentes (Marigold)'
last_update:
  date: 11 January 2024
---

> > Note : clone this [project](https://github.com/marigold-dev/training-security-3.git) for compiling and testing this tutorial

Reviewer: This repo appears to be empty except for a license and readme.

Author: yes, finally I did not pushed code on part 3. A bit too long, but feasible if we have more time and request from users

One of the challenges of designing and deploying a smart contract is to define the roles and permissions of the different parties involved in the execution of the contract. The question is who can do what?

- Who can create, modify, or terminate the contract?
  - Creation: A smart contract is generally deployed by a DevOps person or a CI pipeline. Apart from knowing the creator's address and the fees he paid, there is no critical event that can appear at this moment. The only hack here would be to impersonate another company's smart contract, as discussed in the introduction of this training

>   - Update: By design, a smart contract code cannot be modified because it is immutable. There is an exception below in the Chapter 2

Reviewer: ```suggestion
  - Update: By design, a smart contract code cannot be modified because it is immutable. However, lambdas are an exception, as described in the next section.
```

## Lambda, mutable code and dynamic entrypoints

A smart contract is immutable but not the storage that represents the mutable state.
A common way to change the behavior of a smart contract is to store some mutable code inside its storage and then execute it from the smart contract itself.

This feature presents a **high risk and breaks the trust** you can have in the execution of the contract. If the governance is not clear and transparent, an administrator can potentially push any function and drain your funds from this contract. It is recommended to read carefully the code to see which kind of action can be done through this lambda. A lambda can access the contract's state and call other functions.
For example, if the lambda is called in a function to update a static data configuration that has few impacts, then it is not necessarily dangerous. But if the lambda can create an operation that can be executed or is returning false information to fool the user over a transaction, then it is a red flag

> > Other technics exist to update a dapp like the proxy pattern. It does not change the code of the smart contract, but deployed a new version of it and a proxy contract will redirect the user transaction to this new contract. The risk is located on the proxy contract. If this contract is hacked and badly protected, anyone can deploy a malicious contract and redirect to it

Reviewer: ```suggestion
> Other techniques exist to update a dApp, like the proxy pattern. It does not change the code of the smart contract, but deployed a new version of it and a proxy contract will redirect the user transaction to this new contract. The risk is located on the proxy contract. If this contract is hacked and badly protected, anyone can deploy a malicious contract and redirect to it
```

> [An excellent tutorial here about smart contract upgrades using Lambda and proxies](https://github.com/marigold-dev/training-dapp-4)

Reviewer: ```suggestion
For an example of upgrading smart contracts with lambdas and proxies, see [Create your minimum dapp on Tezos](../dapp).
```

## (Trustable) oracles

> Blockchain oracles are third-party services that provide smart contracts with external information1. They serve as bridges between blockchains and the outside world, allowing smart contracts to access off-chain data. Oracles verify, query, and authenticate external data sources, and transmit any valuable data.

Reviewer: ```suggestion
Blockchain oracles are third-party services that provide smart contracts with external information. They serve as bridges between blockchains and the outside world, allowing smart contracts to access off-chain data. Oracles verify, query, and authenticate external data sources, and transmit any valuable data.
```
